src/liquid/match_visitor.cr

```diff
@@ -0,0 +1,59 @@
+require "./blocks"
+
+module Liquid
+  class MatchVisitor < Visitor
+    def visit(node : Node) : String
+      return node.children.map { |e| self.visit e }.join
+    end
+
+    def visit(node : Block::RawNode) : String
+      return node.content
+    end
+
+    def visit(node : ExpressionNode)
+      return ".*"
+    end
+
+    def visit(node : Conditional) : String
+      child_nodes = node.children.map { |e| self.visit e }
+
+      if elsif_arr = node.elsif
+        elsif_arr.each do |alt|
+          child_nodes << self.visit alt
+        end
+      end
+
+      if else_arr = node.else
+        child_nodes << self.visit else_arr
+      else
+        child_nodes << ""
+      end
+
+      return "(#{child_nodes.join("|")})"
+    end
+
+    def visit(node : Case)
+      child_nodes = [] of String
+
+      if when_arr = node.when
+        when_arr.each do |when_node|
+          child_nodes << self.visit when_node
+        end
+      end
+
+      if else_arr = node.else
+        child_nodes << self.visit else_arr
+      else
+        child_nodes << ""
+      end
+
+      return "(#{child_nodes.join("|")})"
```

Comment:
Why add these `(` and `|` to result string?

What if the Visitor receives the string to be matched in the constructor and `visit` method returned a `Bool` if their contents matches? So we could save some memory allocations and also stop visiting all nodes once we find some node that matches the string.

Reply:
`|` is basically a regex character for OR, while the brackets contain it. It basically creates a regex version of an if/elsif/else statement.

Reply:
got it. But I still believe that the regex must be "the needle", not the "hay stack".